lebedigital/openbis/expstep.py

```diff
@@ -0,0 +1,531 @@
+        """
+        # TYPE CHECKING
+        conv_dict = {
+            'BOOLEAN': bool,
+            'CONTROLLEDVOCABULARY': str,
+            'DATE': str,
+            'HYPERLINK': str,
+            'INTEGER': int,
+            'MATERIAL': str,
+            'MULTILINE_VARCHAR': str,
+            'OBJECT': str,
+            'REAL': float,
+            'TIMESTAMP': str,
+            'VARCHAR': str,
+            'XML': str,
+            'SAMPLE': str
+        }
+        logging.debug('TYPE CHECKING OF THE SAMPLE')
+        logging.debug('Setting up comparison dict')
+        types_df = o.get_sample_properties(self.type)
+        types_dict = dict(zip(types_df.code, types_df.dataType))
+        logging.debug(types_dict)
+        types_dict: dict = {k.lower(): conv_dict[v] for k, v in types_dict.items()}
+
+        # Check if all keys are a subset of all possible keys
+        logging.debug('Checking if keys are subset of defined parameters')
+        if set(self.metadata.keys()).issubset(set(types_dict.keys())):
+            for key, val in self.metadata.items():
+                # Check if all values have the correct data type
+                if not isinstance(val, types_dict[key]):
+                    raise ValueError(
+                        f'Type Checker: Type Mismatch.'
+                        f'Entry "{key}" is of Type {type(val)} and should be {types_dict[key]}')
+                    # Check if float values are not NaN
+                if isinstance(val, float) and isnan(val):
+                    raise ValueError(
+                        f'Type Checker: NaN values are not accepted by Openbis.'
+                        f'Entry "{key} has a value "{val}".')
+        else:
+            # Keys in metadata and not in object definition -> raises an error for every such key
+            for key in self.metadata.keys():
+                if key.lower() not in types_dict.keys():
+                    raise ValueError(
+                        f'Type Checker: Unexpected parameter.'
+                        f'Key "{key}" is not in {self.type} defined metadata params')
+
+        # Checking if all values are defined
+        needed_attrs = ['name', 'type', 'collection', 'space', 'project']
+        for attr, value in self.__dict__.items():
+            if attr in needed_attrs:
+                if not value:
+                    raise ValueError(
+                        f'Type Checker: Undeclared Attribute.'
+                        f'Attribute "{attr}" has not been delcared')
+
+        # Printing warnings if parents are empty, does not break the function
+        if not self.parents:
+            logging.warning(
+                'Type Checker: Parents attribute undeclared')
+
+        # If you got to this line no game breaking bugs were found
+        logging.info('Type Checker: Types are correctly set')
+
+    @staticmethod
+    def load_sample(o: Interbis, sample_identifier: str):
```

Comment:
Wouldnt't that be a method that belongs to the openBIS instance, i.e. an instance of the derived class from Interbis that has the function load_sample and returns an object. At some later stage this could be proposed as an improvement to Interbis

Reply:
If it was under the Interbis class then I would also need to have the ExpStep class as a dependency for Interbis and I think it would break the Idea we had of separating them from each other. It would make more sense to have it as an Interbis method though.